libbeat/idxmgmt/lifecycle/config.go

```diff
@@ -57,18 +73,46 @@ var DefaultPolicy = mapstr.M{
 	},
 }
 
+// DefaultDSLPolicy defines the default policy to be used for DSL if
+// no custom policy is configured
+var DefaultDSLPolicy = mapstr.M{
+	"data_retention": "7d",
+}
+
 // Validate verifies that expected config options are given and valid
 func (cfg *Config) Validate() error {
```

Comment:
Agreed, I think we should stick to the existing Beats configuration paradigm of doing exactly what we are told to do and failing when told to do conflicting things.

This prevents someone writing a single configuration file that works for both stateful and serverless, but a user is already going to have to template or change the hostname so this just add one extra thing to modify. This is similar to how we support multiple outputs, that is we don't and require two separate configuration files instead of inferring which one we connect to.

I want to avoid starting down the path of the Beat figuring out which parameters to use or ignore based on what it is connected to. This will ultimately make the configuration much harder to debug in the long term as sections of it are conditionally applied based on the runtime environment without the conditions being in the configuration itself.

Reply:
So, did some poking around, and I don't think that `Validate()` function is ever called. However, `NewESClientHandler` will check to see if both ILM and DSL are enabled, and will return an error.

Reply:
`Validate()` should be being called when you unpack a lifecycle config.

Reply:
So something like the following in `config.go`

```
func (c *LifecycleConfig) Validate() error {
	if c.ILM.Enabled && c.DSL.Enabled {
		return fmt.Errorf("cannot have both ILM & DSL enabled at same time")
	}
	return nil
}
```

and then tests like:

```
func TestValidateLifecycleConfig(t *testing.T) {
	tests := map[string]struct {
		input                   []byte
		expectedValidationError string
	}{
		"blank": {input: []byte(""),
			expectedValidationError: "",
		},
		"bothEnabled": {input: []byte(`
setup.ilm:
    enabled: true
setup.dsl:
    enabled: true
`),
			expectedValidationError: "cannot have both ILM & DSL enabled at same time accessing config",
		},
	}
	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			cfg, err := yaml.NewConfig(tc.input)
			require.NoError(t, err)
			config := LifecycleConfig{}
			err = cfg.Unpack(&config)
			if tc.expectedValidationError != "" {
				require.Errorf(t, err, "Config was: %v", config)
				require.Equal(t, tc.expectedValidationError, err.Error())
			} else {
				require.NoError(t, err)
			}
		})
	}
}
```